pkg/disttask/framework/dispatcher/dispatcher.go

```diff
@@ -695,20 +637,46 @@ func (d *BaseDispatcher) dispatchSubTask(
 	for i := range serverNodes {
 		d.TaskNodes[i] = disttaskutil.GenerateExecID(serverNodes[i].IP, serverNodes[i].Port)
 	}
+	var size uint64
 	subTasks := make([]*proto.Subtask, 0, len(metas))
 	for i, meta := range metas {
 		// we assign the subtask to the instance in a round-robin way.
 		// TODO: assign the subtask to the instance according to the system load of each nodes
 		pos := i % len(serverNodes)
 		instanceID := disttaskutil.GenerateExecID(serverNodes[pos].IP, serverNodes[pos].Port)
 		logutil.Logger(d.logCtx).Debug("create subtasks", zap.String("instanceID", instanceID))
-		subTasks = append(subTasks, proto.NewSubtask(subtaskStep, d.Task.ID, d.Task.Type, instanceID, d.Task.Concurrency, meta))
+		subTasks = append(subTasks, proto.NewSubtask(
+			subtaskStep, d.Task.ID, d.Task.Type, instanceID, d.Task.Concurrency, meta, i+1))
+
+		size += uint64(len(meta))
 	}
 	failpoint.Inject("cancelBeforeUpdateTask", func() {
-		_ = d.updateTask(proto.TaskStateCancelling, subTasks, RetrySQLTimes)
+		_ = d.taskMgr.CancelTask(d.ctx, d.Task.ID)
 	})
 
-	return d.updateTask(d.Task.State, subTasks, RetrySQLTimes)
+	// as other fields and generated key and index KV takes space too, we limit
+	// the size of subtasks to 80% of the transaction limit.
+	limit := max(uint64(float64(kv.TxnTotalSizeLimit.Load())*0.8), 1)
```

Comment:
uint64(float64(kv.TxnTotalSizeLimit.Load())*0.8) and 1， seems the first parameter value always bigger than 1？

Reply:
there's only a upper limit check, min value of `txn-total-size-limit` is 0